### PR TITLE
make the slow log configurable for the time lock server

### PR DESCRIPTION
**Pull request by hsaraogi**
**Goals (and why)**: Slow lock logging should be enabled or disabled by the timelock server configuration and the trigger time must be configurable too.

**Implementation Description (bullets)**:
1. LockServerOptions has a slowLogTriggerMillis instead of boolean. If the value is positive the logging is  enabled for all locks that take more than that time for the response. Logging is disabled for the LockServer by default.
2. This value is configurable for the lock server and is set to 10s by default.

**Concerns (what feedback would you like?)**: All feedback is welcome.

**Where should we start reviewing?**: LockServerImpl

**Priority (whenever / two weeks / yesterday)**: tomorrow.

### CODE REVIEW

File: lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java [hunk 1]
```diff
@@ -471,20 +469,24 @@ private void tryLocks(LockClient client, LockRequest request, BlockingMode block
...
-        if (slowLogEnabled) {
+    private void logSlowLockAcquisition(String lockId, LockClient currentHolder, long durationMillis) {
+        if (isSlowLogEnabled() && durationMillis >= slowLogTriggerMillis) {
             SlowLockLogger.logger.info("Blocked for {} ms to acquire lock {} {}.",
-                    duration,
+                    durationMillis,
                     lockId,
                     currentHolder == null ? "successfully" : "unsuccessfully");
-        } else {
+        } else if (durationMillis > 100) {
```
Comment (gsheasby) [inline lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java]: You could extract the magic number here. Why is this threshold 100x lower than that on the default TimeLock configuration below?

Comment (gsheasby) [inline lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java]: nit: also, you have `>` here and `>=` above. Shouldn't these be consistent?

File: lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java [hunk 2]
```diff
@@ -471,20 +469,24 @@ private void tryLocks(LockClient client, LockRequest request, BlockingMode block
         }
     }
 
-    private void logSlowLockAcquisition(String lockId, LockClient currentHolder, long duration) {
-        if (slowLogEnabled) {
+    private void logSlowLockAcquisition(String lockId, LockClient currentHolder, long durationMillis) {
```
Comment (gsheasby) [inline lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java]: There are a few forks and flows of control here. Is it time to extract some little testable helper for our slow logging?

File: lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java (hunk 1 above)
Comment (hsaraogi) [inline lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java]: This was in the interest of preserving the bebug logging functionality, see line 454.

File: docs/source/release_notes/release-notes.rst [hunk 3]
```diff
@@ -47,7 +47,8 @@ develop
            (`Pull Request <https://github.com/palantir/atlasdb/pull/1783>`__)
 
     *    - |improved|
-         - Any lock requests that take more than ``100ms`` to receive a response are now logged in the ``SlowLockLogger`` logger.
+         - The lock server can be configured with an option to enable logging if the request that take more than a given time. Specifically, the timelock server also takes a configuration parameter ``slowLockLogTriggerMillis``
+           and any lock server requests that take more than 10 are now logged in the ``SlowLockLogger`` logger.
```
Comment (gsheasby) [inline docs/source/release_notes/release-notes.rst]: more than 10 what? :)

File: lock-impl/src/test/java/com/palantir/lock/impl/LockServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @Test
+    public void slowLogShouldBeEnabled() {
+        assertThat(lockService.isSlowLogEnabled(), is(true));
+    }
+
+    @Test
+    public void slowLogShouldBeLogged() {
+        lockService.logSlowLockAcquisition("test_lockId", LockClient.ANONYMOUS, 10);
```
Comment (gsheasby) [inline lock-impl/src/test/java/com/palantir/lock/impl/LockServiceImplTest.java]: This doesn't seem to actually test anything.

File: docs/source/release_notes/release-notes.rst [hunk 5]
```diff
@@ -47,9 +47,10 @@ develop
            (`Pull Request <https://github.com/palantir/atlasdb/pull/1783>`__)
 
     *    - |improved|
-         - The lock server can be configured with an option to enable logging if the request that take more than a given time. Specifically, the timelock server also takes a configuration parameter ``slowLockLogTriggerMillis``
-           and any lock server requests that take more than 10 are now logged in the ``SlowLockLogger`` logger.
-           (`Pull Request <https://github.com/palantir/atlasdb/pull/1769>`__)
+         - The lock server can be configured with an option to enable logging to the ``SlowLockLogger`` logger if the request takes more than a given time to be processed on the lock server. Specifically, the
```
Comment (gsheasby) [inline docs/source/release_notes/release-notes.rst]: very pedantic nit: prefer one sentence per line in docs and release notes, unless the sentences are very short.

File: lock-impl/build.gradle [hunk 6]
```diff
@@ -7,5 +7,6 @@ dependencies {
   compile 'com.palantir.patches.sourceforge:trove3:' + libVersions.trove
   compile group: 'com.palantir.remoting1', name: 'tracing'
   compile group: 'joda-time', name: 'joda-time'
+  testCompile group: 'uk.org.lidalia', name: 'slf4j-test', version: '1.0.0'
```
Comment (gsheasby) [inline lock-impl/build.gradle]: Did you run `generateLock saveLock`?

File: docs/source/release_notes/release-notes.rst [hunk 7]
```diff
@@ -47,8 +47,10 @@ develop
            (`Pull Request <https://github.com/palantir/atlasdb/pull/1783>`__)
 
     *    - |improved|
-         - Any lock requests that take more than ``100ms`` to receive a response are now logged in the ``SlowLockLogger`` logger.
-           (`Pull Request <https://github.com/palantir/atlasdb/pull/1769>`__)
+         - The lock server can be configured with an option to enable logging to the ``SlowLockLogger`` logger if the request takes more than a given time to be processed on the lock server. Specifically, the
```
Comment (gsheasby) [inline docs/source/release_notes/release-notes.rst]: It should be pointed out that this is now enabled by default.

File: lock-impl/src/test/java/com/palantir/lock/impl/LockServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    public void clearLoggers() {
+        TestLoggerFactory.clear();
+    }
+
+    protected LockService getLockService() {
+        return SerializingProxy.newProxyInstance(LockService.class, lockService);
+    }
+
+    @Test
+    public void slowLogShouldBeEnabled() {
```
Comment (gsheasby) [inline lock-impl/src/test/java/com/palantir/lock/impl/LockServiceImplTest.java]: extra test: slow log should be disabled when trigger time is zero

File: lock-impl/src/test/java/com/palantir/lock/impl/LockServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        assertThat(lockService.isSlowLogEnabled(), is(true));
+    }
+
+    @Test
+    public void slowLogShouldBeLoggedWhenLockResponseIsSlowButShouldNotBeLoggedAtDebug() {
```
Comment (gsheasby) [inline lock-impl/src/test/java/com/palantir/lock/impl/LockServiceImplTest.java]: extra test: debug log should be logged if response time is slow but slow log is disabled.

File: timelock-server/src/main/java/com/palantir/atlasdb/timelock/TimeLockServer.java [hunk 10]
```diff
@@ -43,7 +43,8 @@ default void onStartupFailure() {
      * Creates timestamp and lock services for the given client. It is expected that for each client there should
      * only be (up to) one active timestamp service, and one active lock service at any time.
      * @param client Client namespace to create the services for
+     * @param slowLogTriggerMillis response time for lock requests that triggers slow lock logging
      * @return Invalidating timestamp and lock services
      */
-    TimeLockServices createInvalidatingTimeLockServices(String client);
+    TimeLockServices createInvalidatingTimeLockServices(String client, long slowLogTriggerMillis);
```
Comment (gsheasby) [inline timelock-server/src/main/java/com/palantir/atlasdb/timelock/TimeLockServer.java]: Is this a `DEV-BREAK`?

File: lock-impl/build.gradle (hunk 6 above)
Comment (hsaraogi) [inline lock-impl/build.gradle]: Thats only for non-test dependencies I believe.

File: timelock-server/src/main/java/com/palantir/atlasdb/timelock/TimeLockServer.java (hunk 10 above)
Comment (hsaraogi) [inline timelock-server/src/main/java/com/palantir/atlasdb/timelock/TimeLockServer.java]: Only if someone implements `TimelockServer`. IMO, we dont need to flag this.

Comment (gsheasby) [inline timelock-server/src/main/java/com/palantir/atlasdb/timelock/TimeLockServer.java]: OK. Our autobumps will catch it anyway :)

File: lock-impl/build.gradle (hunk 6 above)
Comment (gsheasby) [inline lock-impl/build.gradle]: You're right, I also added a test dependency today.

File: lock-impl/src/test/java/com/palantir/lock/impl/LockServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+import com.palantir.lock.LockServerOptions;
+
+import uk.org.lidalia.slf4jext.Level;
+import uk.org.lidalia.slf4jtest.TestLogger;
+import uk.org.lidalia.slf4jtest.TestLoggerFactory;
+
+public final class LockServiceImplTest {
+    public static final long SLOW_LOG_TRIGGER_MILLIS = LockServiceImpl.DEBUG_SLOW_LOG_TRIGGER_MILLIS + 10;
+    private static final String TEST_LOCKID = "test_lockId";
+    private static final LockServiceImpl lockServiceWithSlowLogEnabled = createLockServiceWithSlowLogEnabled(true);
```
Comment (gsheasby) [inline lock-impl/src/test/java/com/palantir/lock/impl/LockServiceImplTest.java]: This is flaky - state will be retained by the static LockServiceImpl classes, resulting in more things being logged, and the really-long-name test breaking (see circle build).

**Pull request merged**